### ISSUE TITLE
tf.gfile.Glob returns wrong values on Windows systems

### ISSUE THREAD

**Issue by canut** [tensorflow/tensorflow#19378]
### System information
Environment:
- **Python 3.6.2**
- **Tensorflow 'v1.8.0-0-g93bc2e2072' 1.8.0**
- **OS : Windows 7**

- **Exact command to reproduce**:

```
 > ls -R
.:
example/

./example:
subdir/

./example/subdir:
file1  file2


> python
Python 3.6.2 (v3.6.2:5fd33b5, Jul  8 2017, 04:57:36) [MSC v.1900 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import tensorflow as tf
>>> tf.gfile.Glob("example/*")
['example\\subdir', 'example\\subdir\\file1', 'example\\subdir\\file2']

```

### Describe the problem
Behaviour of _tf.gfile.Glob_ is different on Windows and Linux-based (POSIX) systems.
On Windows, the result includes all files and sub-directories of the matching pattern (=> recursive)
On Linux, the result includes only files and directories matched by the pattern (=> not recursive)

As an example, given the directory structure above, I have the following result:

* Windows
```python
>>> tf.gfile.Glob("example/*")
['example\\subdir', 'example\\subdir\\file1', 'example\\subdir\\file2']
```

* Linux-based systems
```python
>>> tf.gfile.Glob("example/*")
['example/subdir']
```

The result returned on Windows platform is clearly wrong.

This has an impact on my application because I do something like `[my_list]=tf.gfile.Glob(pattern)` and it doesn't work on Windows whereas it works on Linux.
Error on Windows is `ValueError: too many values to unpack `

### Source code
I tracked down the issue in TF code and it is related to how _GetMatchingPaths_ works. See difference between [Windows implementation](https://github.com/tensorflow/tensorflow/blob/e7f158858479400f17a1b6351e9827e3aa83e7ff/tensorflow/core/platform/windows/windows_file_system.cc) and [POSIX implementation](https://github.com/tensorflow/tensorflow/blob/e7f158858479400f17a1b6351e9827e3aa83e7ff/tensorflow/core/platform/posix/posix_file_system.cc).
Both of them refer to [internal::GetMatchingPaths](https://github.com/tensorflow/tensorflow/blob/6ba9573a702ea9e1290517cce4e5f73a14552ad9/tensorflow/core/platform/file_system_helper.cc) which seems to be the root cause of the issue.

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory

**Comment by canut** [comment]
Sure. Please find below the details.

Have I written custom code : No
OS Platform and Distribution : Windows 7
TensorFlow installed from : pip 
TensorFlow version: Tensorflow 'v1.8.0-0-g93bc2e2072' 1.8.0
Bazel version : N/A
CUDA/cuDNN version : N/A
GPU model and memory : N/A

**Comment by karmel** [comment]
@rohan100jain - do you have any experience with gfile.Glob on Windows?

**Comment by tensorflowbutler** [comment]
Nagging Assignee @karmel: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @karmel: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by Cibifang** [comment]
I checked the code and maybe found the reason.

This is because Env::MatchPath (What the tf.gfile.Glob will last use) has different behavior on Windows and Linux.
There is a related issue which was already closed 
[#15501](https://github.com/tensorflow/tensorflow/issues/15501)

Maybe @snnn and @jart will have some suggestions?

**Comment by amitport** [comment]
It's been two years since the last update. Do `contributions welcome`-issues return to the backlog at some point?

**Comment by snnn** [comment]
Sorry I couldn't continue working on this. TF's source tree has changed a lot in the past 2 years, I'm not familiar with the latest.

**Comment by mihaimaruseac** [comment]
Can you try with the 2.4 release please?

**Comment by mihaimaruseac** [comment]
Should be fixed by the 2.4 release. If it isn't, please reopen.
